lib/recommendations.js

```diff
@@ -0,0 +1,171 @@
+const select_recommendation_resource_type = query.bind(query, fs.readFileSync('./sql/select_recommendation_resource_type.sql').toString('utf8'), null);
+
+/**
+ * Returns an array of available resource types for recommendations
+ */
+async function get_resource_types(pg_pool) {
+  return select_recommendation_resource_types(pg_pool, []);
+}
+
+async function check_resource_type_exists(pg_pool, resource_type_key) {
+  const resource_type = (await select_recommendation_resource_type(pg_pool, [resource_type_key]));
+  if (resource_type.length === 0) {
+    throw new httph.NotFoundError(`No recommendation type was found with key "${resource_type_key}"`);
+  }
+  return resource_type[0];
+}
+
+async function check_recommendation_exists(pg_pool, app_uuid, service, resource_type, action) {
+  const recommendations = await select_recommendation_by_primary_key(pg_pool, [app_uuid, service, resource_type, action]);
+  if (recommendations.length === 0) {
+    throw new httph.NotFoundError(`No recommendation was found for ${service} with action ${action} on app ${app_uuid} and resource type ${resource_type}.`);
+  }
+  return recommendations[0];
+}
+
+// Stub for fetching recommendations
+// async function get() {}
+
+// Stub for listing recommendations
+// async function list() {}
+
+async function check_payload(pg_pool, payload) {
+  assert.ok(payload.service, 'Missing service property - A originating service must be specified when creating a new recommendation');
+  assert.ok(payload.action, 'Missing action property - An action must be specified when creating a new recommendation');
+  assert.ok(payload.details, 'Missing details property - Details must be provided when creating a new recommendation');
+  assert.ok(payload.resource_type.actions.split(',').some((x) => x === payload.action),
+    'Invalid action - Specified action is not valid for the given resource type.');
+  // Check any necessary payload details here
+  assert.ok(payload.details.description, 'Missing description - A human readable description must be provided in the details for a new recommendation');
+}
+
+/**
+ * Create new recommendation, or update existing recommendation if one exists for the given app, service, and resource type
+ * @param {*} pg_pool - Database connection pool
+ * @param {*} app_key - App identifier
+ * @param {*} service - Originating service
+ * @param {*} resource_type - Type of resource (e.g. formation) - should be UUID
+ * @param {*} details - JSON object containing recommendation details (such as a human readable description)
+ */
+async function create_or_update(pg_pool, app_uuid, service, resource_type, action, details) {
+  // We only want there to be one cached recommendation per app/service/resource_type/action so upon
+  // create if it already exists then update instead
+
+  let recommendation = null;
+  try {
+    recommendation = await check_recommendation_exists(pg_pool, app_uuid, service, resource_type, action);
```

Comment:
How would you feel about having `check_recommendation_exists` return something falsey if a recommendation is not found and then doing a conditional instead of a try/catch? I guess I don't love the pattern of using try/catch for conditionals, especially when we can control whether an error was thrown or not.

I'm good with whichever way you decide, just wanted to get your take on it.

Reply:
This is following a pattern that is commonly used in the code - when you call `const recommendation = await check_recommendation_exists(...)` in any http handler, and the given recommendation does not exist, a 404 response will automatically be sent as a response.

For example:

https://github.com/akkeris/controller-api/blob/220b6ffab444d073532a7d7606b484cc88d9e8b6/lib/common.js#L318-L325

https://github.com/akkeris/controller-api/blob/220b6ffab444d073532a7d7606b484cc88d9e8b6/lib/addons.js#L172-L174

When `check_app_exists` is called from `http_create` for an addon (or in any other http handler) and the app doesn't exist, an error response will automatically be returned to the user without having to do anything else.

I'm doing a `try catch` around this function here because I'm essentially reusing this function that is going to be primarily used as part of a user-called http handler. I'm just using it in a place where I don't want the error to be passed to the user. I could have written another function to query the database and returned a boolean but I figured it was easier and less code to just wrap that same function in a `try catch`

Reply:
Sounds good to me. Thanks for the detailed response.